### PR TITLE
BUG: Fix concat series loss of timezone

### LINKED ISSUE #23816: Concat Series[datetimetz] & Series[category] loses timezone
```python

In [8]: a = pd.Series(pd.date_range('20170101', periods=4, tz='US/Pacific'))

In [9]: b = pd.Series(['a', 'b'], dtype='category')

In [10]: pd.concat([a, b], ignore_index=True)[0]
Out[10]: Timestamp('2017-01-01 08:00:00')

```

Out[10]  should have a tz like the following

```python
In [11]: pd.concat([a, b.astype(object)], ignore_index=True)[0]
Out[11]: Timestamp('2017-01-01 00:00:00-0800', tz='US/Pacific', freq='D')
```

### PR DESCRIPTION
**Pull request by evangelinehl**
Closes #23816

### CODE REVIEW

**Comment by TomAugspurger** [comment]
Can you add “closes 
https://github.com/pandas-dev/pandas/issues/23816” to the original post?

We also need a test (the one from the issue is fine) and a release note.

File: pandas/core/dtypes/concat.py [hunk 1]
```diff
@@ -193,7 +193,8 @@ def _concat_categorical(to_concat, axis=0):
 
     def _concat_asobject(to_concat):
         to_concat = [x.get_values() if is_categorical_dtype(x.dtype)
-                     else np.asarray(x).ravel() for x in to_concat]
+                     else np.asarray(x).ravel() if not is_datetime64tz_dtype(x)
```
Comment (jreback) [inline pandas/core/dtypes/concat.py]: this is not the way
_concat_conpat already handles this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by TomAugspurger** [comment]
_concat_compat isn’t called from categorical. Maybe it could be.


________________________________

**Comment by evangelineliu** [comment]
Hi, so should we be calling from concat_compat?

**Comment by jakezimmer** [comment]
Just wanted to test something here, I'm decently sure that these tests will fail.

File: pandas/core/dtypes/concat.py (hunk 1 above)
Comment (jakezimmer) [reply pandas/core/dtypes/concat.py]: @jreback after closer examination, I'm not sure if  _concat_compat actually already handles this. If the type is also a category, it simply calls _concat_categorical which then summons _concat_asobject. I think that therefore the change to the code should happen inside _concat_categorical or _concat_asobject.

**Review by jreback — CHANGES_REQUESTED** [review]
this needs a test first thing

**Comment by evangelineliu** [comment]
Sorry, still trying to figure out why the dimensions would be different for the arrays passed to np.concatenate. In your experience why does this usually happen?

**Comment by jreback** [comment]
> Sorry, still trying to figure out why the dimensions would be different for the arrays passed to np.concatenate. In your experience why does this usually happen?

datetimes with timezones are backed by a DatetimeIndex itself which is only 1-D object. datetimes that are naive are 2-D blocks internaly.

**Comment by evangelineliu** [comment]
Taking a look at concat_compat, I agree with @jakezimmer that I'm not super sure if it actually takes care of the if checks that we tried to delete.

**Comment by jakezimmer** [comment]
@jreback Just wondering if you saw my comments above on your initial change request about my concern over _concat_compat not being able to handle this

File: pandas/core/dtypes/concat.py [hunk 2]
```diff
@@ -23,7 +23,6 @@ def get_dtype_kinds(l):
     Parameters
     ----------
     l : list of arrays
-
```
Comment (TomAugspurger) [inline pandas/core/dtypes/concat.py]: These blank lines will need to be restored.

File: pandas/core/dtypes/concat.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/dtypes/concat.py]: I guess I'm not sure why we aren't doing

```python
to_concat = [np.asarray(x.astype(object)) for x in to_concat]
```

The function name is `_concat_asobject`, so let's convert to object and concat them. Can you see if that works?

Please write the test first though, and ensure that the test fails with master. Then implement the fix and make sure it works.

**Comment by TomAugspurger** [comment]
FYI, here's a test

```diff
diff --git a/pandas/tests/reshape/test_concat.py b/pandas/tests/reshape/test_concat.py
index 4113fb7f0..bb002f151 100644
--- a/pandas/tests/reshape/test_concat.py
+++ b/pandas/tests/reshape/test_concat.py
@@ -2552,3 +2552,16 @@ def test_concat_series_name_npscalar_tuple(s1name, s2name):
     result = pd.concat([s1, s2])
     expected = pd.Series({'a': 1, 'b': 2, 'c': 5, 'd': 6})
     tm.assert_series_equal(result, expected)
+
+
+def test_concat_categorical_tz():
+    # GH-23816
+    a = pd.Series(pd.date_range('2017-01-01', periods=2, tz='US/Pacific'))
+    b = pd.Series(['a', 'b'], dtype='category')
+    result = pd.concat([a, b], ignore_index=True)
+    expected = pd.Series([
+        pd.Timestamp('2017-01-01', tz="US/Pacific"),
+        pd.Timestamp('2017-01-02', tz="US/Pacific"),
+        'a', 'b'
+    ])
+    tm.assert_series_equal(result, expected)
```

**Comment by evangelineliu** [comment]
Is this a test for the new output? Because I tried what you said earlier already but it didn't build

File: pandas/core/dtypes/concat.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/dtypes/concat.py]: Seems like we have some inconsistencies in `concat` that's breaking this approach. IIRC there's an outstanding PR fixing some of these up.

I was going by the loose rule of "different dtypes means the result type is object". But things like 

```python
In [1]: import pandas as pd; import numpy as np; import pickle

In [2]: a = pd.Series([1, 2], dtype='category')

In [3]: b = pd.Series([1, None], dtype='category')

In [4]: pd.concat([a, b], ignore_index=True)
Out[4]:
0    1.0
1    2.0
2    1.0
3    NaN
dtype: float64
```

mess that up. Not sure what's best here...

**Comment by TomAugspurger** [comment]
That's the expected output from whatever ends up fixing the original issue.

**Comment by evangelineliu** [comment]
Should we add in the test case? I'm a little bit confused about what's happening exactly with the tests

**Comment by TomAugspurger** [comment]
Yes. That test is asserting the correct behavior when concatenating a Categorical and timezone-aware series. The expected output is an object-dtype series.

**Comment by jreback** [comment]
@evangelineliu pushed a small cleanup. pls add a whatsnew note (reshaping bug fixes). ping on green.

**Comment by jakezimmer** [comment]
Hi @jreback, I noticed that the fix up [commit](https://github.com/pandas-dev/pandas/pull/24027/commits/c7dcdb445506a0a00fbd9363d6551b0a44d5c259) broke some linting as you can see [here](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=4657)

**Comment by TomAugspurger** [comment]
I'm working soon that in https://github.com/pandas-dev/pandas/pull/24092

**Comment by jakezimmer** [comment]
@jreback all tests passed!

**Review by jreback — CHANGES_REQUESTED** [review]
ping on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks ping on green.

**Comment by jakezimmer** [comment]
@jreback do you know what would cause pandas azure pipelines job to fail like this on the "before install" task. It looks like it couldn't download anaconda. The only change was removing that whitespace and it has never happened in previous commits. Is there a way to try that test?

You can see the error here under Linux py36_locale_slow:
https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=4780

**Comment by jreback** [comment]
thanks @jakezimmer 

that pipeline failed on some network timeout.

**Pull request merged**